File: r/llikert/R/http.R
# HTTP transport with the shared bounded retry policy (decision D16):
# retry 429/502/503/504, timeouts and connection failures, up to `max_tries`, with full
# jitter capped at 60 s and `Retry-After` honored; never retry other statuses.

retry_statuses <- c(429L, 502L, 503L, 504L)

llikert_sleep <- function(seconds) {
  (getOption("llikert.sleep") %||% Sys.sleep)(seconds)
}

backoff_seconds <- function(attempt, base = 1, cap = 60) {
  stats::runif(1, 0, min(cap, base * 2^attempt))
}

retry_after_seconds <- function(value) {
  if (is.null(value) || !nzchar(value)) return(NULL)
  if (grepl("^[0-9]+$", value)) return(as.numeric(value))
  when <- suppressWarnings(as.POSIXct(value, format = "%a, %d %b %Y %H:%M:%S", tz = "GMT"))
  if (is.na(when)) return(NULL)
  max(0, as.numeric(difftime(when, Sys.time(), units = "secs")))
}

engine_request <- function(engine, path, body = NULL, auth = TRUE) {
  req <- httr2::request(paste0(engine$url, path))
  req <- httr2::req_headers(req, Accept = "application/json", `User-Agent` = sprintf("llikert-r/%s", utils::packageVersion("llikert")))
  token <- engine_token(engine)
  if (auth && nzchar(token)) req <- httr2::req_headers(req, Authorization = paste("Bearer", token), .redact = "Authorization")
  if (!is.null(engine$scale_up_timeout)) req <- httr2::req_headers(req, `X-Scale-Up-Timeout` = as.character(engine$scale_up_timeout))
  req <- httr2::req_timeout(req, engine$timeout)
  # redirects are not followed, so credentials never reach another host
  req <- httr2::req_options(req, followlocation = 0L)
  req <- httr2::req_error(req, is_error = function(resp) FALSE)
  if (!is.null(body)) {
    req <- httr2::req_body_raw(req, charToRaw(enc2utf8(as.character(to_json(body)))), type = "application/json")
  }
  req
}

# Perform a request; returns the parsed JSON body of a 200 response or signals a condition.
engine_call <- function(engine, path, body = NULL, retry = TRUE, auth = TRUE, call = rlang::caller_env()) {
  tries <- if (retry) engine$max_tries else 1L
  req <- engine_request(engine, path, body, auth)
  for (attempt in seq_len(tries)) {
    resp <- tryCatch(httr2::req_perform(req), httr2_failure = function(e) e, curl_error = function(e) e)
    if (inherits(resp, "condition")) {
      if (attempt < tries) {
        llikert_sleep(backoff_seconds(attempt - 1L))
        next
      }
      llikert_abort("Could not reach the scoring service (connection failure or timeout).", "transport", parent = resp, call = call)
    }
    status <- httr2::resp_status(resp)
    if (status == 200L) {
      proto <- httr2::resp_header(resp, "LLikert-Protocol")
      if (!is.null(proto) && proto != as.character(protocol_version)) {
        llikert_abort("The service speaks a different protocol version.", "protocol_version_mismatch", call = call)
      }
      parsed <- tryCatch(from_json(httr2::resp_body_string(resp, encoding = "UTF-8")), error = function(e) NULL)
      if (is.null(parsed)) llikert_abort("The service returned a non-JSON success response.", "invalid_response", call = call)
      return(parsed)
    }
    if (status %in% retry_statuses && attempt < tries) {
      wait <- retry_after_seconds(httr2::resp_header(resp, "Retry-After"))
      llikert_sleep(if (is.null(wait)) backoff_seconds(attempt - 1L) else min(60, wait))
      next
    }
    signal_response_error(resp, call)
  }
}

signal_response_error <- function(resp, call) {
  status <- httr2::resp_status(resp)
  body <- tryCatch(from_json(httr2::resp_body_string(resp, encoding = "UTF-8")), error = function(e) NULL)
  err <- if (is.list(body)) body$error else NULL
  if (is.list(err) && rlang::is_string(err$code) && rlang::is_string(err$message)) {
    service_error(status, err$code, err$message, err$details, err$request_id, call = call)
  }
  text <- tryCatch(httr2::resp_body_string(resp), error = function(e) "")
  excerpt <- substr(trimws(gsub("\\s+", " ", gsub("<[^>]*>", " ", text))), 1, 200)
  message <- sprintf("HTTP %s without a protocol error body%s", status, if (nzchar(excerpt)) paste0(": ", excerpt) else "")
  service_error(status, "http_error", message, call = call)
}

File: python/src/llikert/_http.py
"""HTTP transport with the shared bounded retry policy (decision D16)."""

from __future__ import annotations

import email.utils
import json
import random
import re
import time
from dataclasses import dataclass, field
from typing import Any, Callable

import httpx

from llikert._errors import AuthenticationError, FingerprintMismatch, PrepareError, ServiceError, ServiceUnavailable, TransportError
from llikert._version import PROTOCOL_VERSION, __version__

RETRY_STATUSES = frozenset({429, 502, 503, 504})


@dataclass(frozen=True)
class RetryPolicy:
    max_tries: int = 6
    base_seconds: float = 1.0
    cap_seconds: float = 60.0
    sleep: Callable[[float], None] = field(default=time.sleep, repr=False, compare=False)

    def backoff(self, attempt: int) -> float:
        """Full jitter: uniform in [0, min(cap, base * 2**attempt)]."""
        return random.uniform(0, min(self.cap_seconds, self.base_seconds * 2**attempt))


def retry_after_seconds(value: str | None) -> float | None:
    if not value:
        return None
    value = value.strip()
    if value.isdigit():
        return float(value)
    try:
        when = email.utils.parsedate_to_datetime(value)
    except (TypeError, ValueError):
        return None
    return max(0.0, when.timestamp() - time.time())


def _excerpt(text: str, limit: int = 200) -> str:
    text = re.sub(r"<[^>]*>", " ", text)
    return re.sub(r"\s+", " ", text).strip()[:limit]


def error_from_response(resp: httpx.Response) -> ServiceError:
    try:
        body = resp.json()
        err = body["error"]
        code, message = str(err["code"]), str(err["message"])
        details, request_id = err.get("details"), err.get("request_id")
    except (ValueError, KeyError, TypeError):
        code = "http_error"
        message = f"HTTP {resp.status_code} without a protocol error body"
        excerpt = _excerpt(resp.text)
        if excerpt:
            message += f": {excerpt}"
        details, request_id = None, None
    status = resp.status_code
    if status in (401, 403):
        cls = AuthenticationError
    elif status == 409:
        cls = FingerprintMismatch
    elif status == 422 and code in {"invalid_response_codes", "invalid_task", "task_exceeds_context", "render_error", "boundary_error"}:
        cls = PrepareError
    elif status in (429, 503):
        cls = ServiceUnavailable
    else:
        cls = ServiceError
    return cls(status, code, message, details, request_id)


class Transport:
    def __init__(
        self,
        url: str,
        token: str | None = None,
        timeout: float = 300.0,
        retry: RetryPolicy = RetryPolicy(),
        scale_up_timeout: int | None = None,
        client: httpx.Client | None = None,
    ):
        if not url:
            raise ValueError("a service URL is required")
        self.url = url.rstrip("/")
        headers = {"User-Agent": f"llikert-python/{__version__}", "Accept": "application/json"}
        if token:
            headers["Authorization"] = f"Bearer {token}"
        if scale_up_timeout is not None:
            headers["X-Scale-Up-Timeout"] = str(int(scale_up_timeout))
        self.retry = retry
        # redirects are not followed, so credentials never reach another host
        self._client = client or httpx.Client(timeout=httpx.Timeout(timeout, connect=min(30.0, timeout)), follow_redirects=False)
        self._headers = headers

    def close(self) -> None:
        self._client.close()

    def request(self, method: str, path: str, body: Any = None, *, retry: bool = True, auth: bool = True) -> Any:
        payload = None
        if body is not None:
            payload = json.dumps(body, ensure_ascii=False, allow_nan=False).encode("utf-8")
        headers = dict(self._headers)
        if not auth:
            headers.pop("Authorization", None)
        if payload is not None:
            headers["Content-Type"] = "application/json"
        tries = self.retry.max_tries if retry else 1
        last_error: Exception | None = None
        for attempt in range(tries):
            try:
                resp = self._client.request(method, self.url + path, content=payload, headers=headers)
            except (httpx.TimeoutException, httpx.TransportError) as exc:
                last_error = TransportError(f"{type(exc).__name__} contacting the scoring service")
                if attempt + 1 < tries:
                    self.retry.sleep(self.retry.backoff(attempt))
                    continue
                raise last_error from None
            if resp.status_code == 200:
                if resp.headers.get("LLikert-Protocol", str(PROTOCOL_VERSION)) != str(PROTOCOL_VERSION):
                    raise ServiceError(200, "protocol_version_mismatch", "the service speaks a different protocol version")
                try:
                    return resp.json()
                except ValueError:
                    raise ServiceError(200, "invalid_response", "the service returned a non-JSON success response") from None
            error = error_from_response(resp)
            if resp.status_code in RETRY_STATUSES and attempt + 1 < tries:
                wait = retry_after_seconds(resp.headers.get("Retry-After"))
                self.retry.sleep(min(self.retry.cap_seconds, wait) if wait is not None else self.retry.backoff(attempt))
                last_error = error
                continue
            raise error
        raise last_error  # pragma: no cover - loop always returns or raises
